### PR TITLE
[AIRFLOW-3701] Google Cloud Vision Product Search operators

### PR DESCRIPTION
**Pull request by sprzedwojski**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-3701/) issues and references them in the PR title.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Implemented the following Cloud Vision Product Search operators:

- CloudVisionProductSetCreateOperator
- CloudVisionProductSetUpdateOperator
- CloudVisionProductSetGetOperator
- CloudVisionProductSetDeleteOperator
- CloudVisionProductCreateOperator
- CloudVisionProductUpdateOperator
- CloudVisionProductGetOperator
- CloudVisionProductDeleteOperator

### Tests

- [ ] My PR adds the following unit tests:

`test_gcp_vision_hook.py`
`test_gcp_vision_operator.py`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by sprzedwojski** [comment]
CC @kaxil, would be great if you could take a look :)

**Comment by feng-tao** [comment]
not to against this pr, but is it possible to move this pr(or other similar pr) to some other plugin repos (e.g. https://github.com/airflow-plugins) ? Airflow release only happens once every few months. If you want to reiterate your code fast and get a new release for your operator bug fix, it would be better put on a separate repo instead Airflow?

Besides not sure of other committers, I personally have no experience on this kinda operators and it would be hard to review as well.

I would like to hear other committers idea on this as well: @ashb @kaxil @bolkedebruin

**Comment by mik-laj** [comment]
@feng-tao I've been thinking about it for a long time. Unfortunately, this is not a simple and easy matter

Short story:
I have been watching this topic for a long time and difficulties have appeared that made this change difficult.
See: https://lists.apache.org/thread.html/b216d209abff1644fe2cd9e2c144b921b7a84ceead1c8ef6c35103cd@%3Cdev.airflow.apache.org%3E

Long story: 
I've been thinking about Airflow's modularity for a long time.  When I started working on Airflow, support for plugins was limited. It was not possible to install the plugin via "pip", so the installation was complicated. In my opinion, this excluded the use of plug-in mechanism in public production solutions.

I found an [old PR](https://github.com/apache/airflow/pull/730), which introduced support for plugins installed via pip. I fell in love with this PR, so i tried to revive the old PR. For this purpose, i wrote message to @kaxil  on Slack for support. He wrote [a comment](https://github.com/apache/airflow/pull/730#issuecomment-443648731) on December 3 in the reaction.

In the meantime, an alternative PRs has appeared. https://github.com/apache/airflow/pull/4412. It was accepted quickly

The discussion about modularity on the dev list has started. During these discussion, it turned out that splitting Airflow into plugins is not such a simple matter.
https://lists.apache.org/thread.html/b216d209abff1644fe2cd9e2c144b921b7a84ceead1c8ef6c35103cd@%3Cdev.airflow.apache.org%3E
@potiuk (I and @sprzedwojski  work in one company) participated in this discussion. We was waiting for community reactions, what will happen next in the matter of accepting AIP-8

If you think that it is worth to start changes on this matter, I invite you to the discussion on the mailing list.

I have the hope that my explanations will be helpful to you.

Thank you very much for your suggestion.

Disclaimer: This is just my personal statement. I described my observations that are public. Other people in my company may think differently. I believe that the presented story allows us to better understand the context.

**Comment by feng-tao** [comment]
If you are talking about that AIP, I am surely +1 on moving those community built hooks/operators etc into other plugins/packages, but just haven't had time to comment on various threads. But I don't agree if the Airflow committers should be the only maintainers to maintain those packages. I know it is a not very simple solution(which one to move out, which one to stay, who should maintain, how to handle dependency etc). 

IMO, It is hard from a user standpoint to decide which one to use, which one is actively maintained and whether it is released in a given Airflow version. In fact, we(Lyft) just rely on the plugin system internally to build various operator/hook which makes our development iteration much faster.

**Comment by sprzedwojski** [comment]
Hi @feng-tao, I think the change you're proposing is a major undertaking and looking through previous discussions on the devlist I see that no consensus has been reached so far as to the exact solution.

I think the idea is worth considering but given that we've been contributing GCP operators this way for half a year now I'd suggest that this PR follow the same pattern.
And if the community decides it's worth moving the operators to other packages/projects we can then refactor all of them.

@kaxil, maybe you could take a look at the PR, given you're previous involvement in PRs with other GCP operators? Many thanks in advance :)

**Comment by ashb** [comment]
-1000 to operators in plugins. They are just python modules.

I'm not against them being separate python modules, just 100% against making them plugins, there is no need.

```
pip install google-airflow-operators
```

and then

```
from google.airflow.operators import SomeVisionyThingOperator
```

Nothing plugin-y needed.

**Comment by OmerJog** [comment]
@feng-tao
It's not that simple to have operators in plugins. operator interact with hook (sometimes more than one).
Hooks are in the core and subject to changes. Sometimes operators can not work without a change that needs to be submitted to the hook.  When you make a change in hook it could break other operators that relay on this hook. it's very hard to keep track of that if the operators are in another project.

Your idea of frequent releases for plugins can work only if the extended functionality is unrelated to Airflow hooks. This is a big if.

File: airflow/contrib/hooks/gcp_vision_hook.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+from airflow import AirflowException
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+
+PARENT_TEMPLATE = 'projects/{}/locations/{}'
+PRODUCT_SET_NAME_TEMPLATE = 'projects/{}/locations/{}/productSets/{}'
+PRODUCT_NAME_TEMPLATE = 'projects/{}/locations/{}/products/{}'
+
+
+# noinspection PyAbstractClass
```
Comment (kaxil) [inline airflow/contrib/hooks/gcp_vision_hook.py]: We should remove any `noinspection` stuff. Let's aim to fix this on BaseOperator level.

**Review by kaxil — APPROVED** [review]
Although I agree with @feng-tao 's argument, as we have not yet decided an approach on splitting out contrib packages into separate repos, let's merge this one. Once an approach has been finalised and work on it is started we can migrate it into its own repo.


@sprzedwojski A minor comment - Can you please resolve it.

File: airflow/contrib/hooks/gcp_vision_hook.py (hunk 1 above)
Comment (sprzedwojski) [reply airflow/contrib/hooks/gcp_vision_hook.py]: Thanks @kaxil, I removed it. Also I've applied small refactoring according to @mik-laj's suggestion (https://github.com/PolideaInternal/airflow/pull/51#discussion_r256013832) and renamed the `get_client()` method in hook to `get_conn()` to be consistent with other hooks.

File: airflow/contrib/example_dags/example_gcp_vision.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+Example Airflow DAG that creates, gets, updates and deletes Products and Product Sets in the Google Cloud
+Vision service in the Google Cloud Platform.
+
+This DAG relies on the following OS environment variables
+
+* GCP_VISION_LOCATION - Google Cloud Platform zone where the instance exists.
+"""
+import os
+
+# [START howto_operator_vision_retry_import]
```
Comment (kaxil) [inline airflow/contrib/example_dags/example_gcp_vision.py:30]: Why do we have this?? We don't use this in docs.

File: airflow/contrib/example_dags/example_gcp_vision.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+from google.api_core.retry import Retry
+# [END howto_operator_vision_retry_import]
+# [START howto_operator_vision_productset_import]
```
Comment (kaxil) [inline airflow/contrib/example_dags/example_gcp_vision.py:33]: Same

File: airflow/contrib/hooks/gcp_vision_hook.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        try:
+            name = response.name
+        except AttributeError as e:
+            raise AirflowException('Unable to get name from response... [{}]\n{}'.format(response, e))
+        match = re.search(r"[^/]*$", name)
+        if match and match.group(0):
+            product_set_id = match.group(0)
+        else:
+            raise AirflowException('Unable to get id from name... [{}]'.format(name))
+        return product_set_id
```
Comment (mik-laj) [inline airflow/contrib/hooks/gcp_vision_hook.py]: If I understand it correctly, you can shorten it to
name = response.get('name')
if '/' not in name:
            raise AirflowException('Unable to get name from response... [{}]\n{}'.format(response, e))
return name.rsplit('/', 1)[1]

File: airflow/contrib/hooks/gcp_vision_hook.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        if not location or not entity_id:
+            # Not enough parameters to construct the name. Trying to use the name from Product / ProductSet.
+            if not explicit_name:
+                raise AirflowException("Unable to determine the {} name. Please either set the name "
+                                       "directly in the {} object or provide the `location` and "
+                                       "`{}` parameters.".format(entity_label, entity_label, entity_id_label))
+        else:
+            template_name = CloudVisionHook._get_entity_name(is_product, project_id, location, entity_id)
+            if not explicit_name:
+                entity.name = template_name
```
Comment (mik-laj) [inline airflow/contrib/hooks/gcp_vision_hook.py]: I think, you should avoid modify objects from user.  He may not know about it and want to use the same object in the loop.
https://softwareengineering.stackexchange.com/questions/245767/is-modifying-an-incoming-parameter-an-antipattern

File: airflow/contrib/example_dags/example_gcp_vision.py (hunk 2 above)
Comment (sprzedwojski) [reply airflow/contrib/example_dags/example_gcp_vision.py:30]: Actually it is used in the docs (example: https://github.com/apache/airflow/pull/4665/files#diff-40e5e831dc73c82a2aba583e5f89f8a8R2035, `operator.rst` file, line 2035).

We wanted to include imports in the "How-to" documentation to allow the user to copy-paste the code from there and use it right away, without the need to refer to the example DAG's source code.

File: airflow/contrib/example_dags/example_gcp_vision.py (hunk 3 above)
Comment (sprzedwojski) [reply airflow/contrib/example_dags/example_gcp_vision.py:33]: Same as above: https://github.com/apache/airflow/pull/4665#discussion_r257611376

**Comment by sprzedwojski** [comment]
Added minor refactoring changes, mostly to `gcp_vision_hook.py`.

**Comment by sprzedwojski** [comment]
Added a minor change in `test_gcp_vision_operator_system.py`: changed the `gcp_key` used to a CloudVision-specific one.

**Comment by sprzedwojski** [comment]
Hi @kaxil, would you be able to find a moment so we could merge this one?

**Comment by kaxil** [comment]
Thanks @sprzedwojski . I am on annual leave, hence have little to no access to the laptop and internet

**Comment by sprzedwojski** [comment]
Thanks @kaxil, I appreciate you taking the time while on vacations. Enjoy your holidays!

**Pull request merged**